app/grandchallenge/algorithms/migrations/0063_algorithminterface_algorithminterfacethroughtable_and_more.py

```diff
@@ -0,0 +1,137 @@
+# Generated by Django 4.2.16 on 2024-11-25 14:40
+import django.db.models.deletion
+from django.core.exceptions import ObjectDoesNotExist
+from django.db import migrations, models
+from django.db.models import Count
+
+
+def create_algorithm_interfaces(apps, _schema_editor):
+    Algorithm = apps.get_model("algorithms", "Algorithm")  # noqa: N806
+    AlgorithmInterface = apps.get_model(  # noqa: N806
+        "algorithms", "AlgorithmInterface"
+    )
+    AlgorithmInterfaceThroughTable = apps.get_model(  # noqa: N806
+        "algorithms", "AlgorithmInterfaceThroughTable"
+    )
+    ComponentInterface = apps.get_model(  # noqa: N806
+        "components", "ComponentInterface"
+    )
+
+    for algorithm in Algorithm.objects.all():
+        inputs = ComponentInterface.objects.filter(
+            pk__in=algorithm.inputs.values_list("pk", flat=True)
+        ).all()
+        outputs = ComponentInterface.objects.filter(
+            pk__in=algorithm.outputs.values_list("pk", flat=True)
+        ).all()
+
+        try:
+            io = AlgorithmInterface.objects.annotate(
+                input_count=Count("inputs", distinct=True),
+                output_count=Count("outputs", distinct=True),
+            ).get(
+                inputs__in=inputs,
```

Comment:
We cannot use the model manager method here because migrations work with historic model instances, and don't have access to custom manager methods or model methods. Is that ok given that this is a one off migration? Or would it be better if I created a function to do this that I import here and in all other places where I need it (i.e. the forms)? A manager method seemed better than a separate utility function to me, but I'm open for other suggestions.

Reply:
Maybe create a method that the manager uses and can also be imported here. Important to reuse and test as I think there is the same bug as I just mentioned in this query too.

Reply:
Model methods have the same issue. i.e. can't be called in a migration. The only alternative I can think of then is to write a standalone function that does this and call that in all places where we need it.

Reply:
Actually, since migrations only work with historical models, I will need to also pass the model to the function. I'm leaving that as an optional argument to the function with a default at the current `AlgorithmInterface` model.

Reply:
Does the data migration also need to go into its own file? I remember seeing problems with combined table and data modifications in the past.

Reply:
In my tests (with an admittedly small sample of algorithms to migrate) it worked in one go. I can separate them out just in case. But out of curiosity: what issues did we encounter in the past with combined table and data modifications?

Reply:
From what I recall the database would refuse to apply it at runtime. This would have been >7 years ago so my memory is hazy. I think it is best to separate them just in case it is a difference in configuration between development and prod, the Django docs recommend keeping them separate:

> Migrations that alter data are usually called “data migrations”; they’re best written as separate migrations, sitting alongside your schema migrations.

https://docs.djangoproject.com/en/5.1/topics/migrations/#data-migrations